asserts/ifacedecls.go

```diff
@@ -220,7 +222,256 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
+	return matcher.result
+}
+
+var (
+	AlwaysMatchAttributes = &AttributeConstraints{matcher: fixedAttrMatcher{nil}}
+	NeverMatchAttributes  = &AttributeConstraints{matcher: fixedAttrMatcher{errors.New("not allowed, never matching")}}
+)
+
 // Check checks whether attrs don't match the constraints.
 func (c *AttributeConstraints) Check(attrs map[string]interface{}) error {
 	return c.matcher.match("", attrs)
 }
+
+func checkMapOrShortcut(context string, v interface{}) (m map[string]interface{}, shortcut bool, err error) {
+	switch x := v.(type) {
+	case map[string]interface{}:
+		return x, false, nil
+	case string:
+		switch x {
+		case "true":
+			return nil, true, nil
+		case "false":
+			return nil, false, nil
+		default:
+			return nil, false, fmt.Errorf("%s is not a map but is set to a shortcut that is not 'true' nor 'false': %q", context, x)
+		}
+	default:
+		return nil, false, fmt.Errorf("%s must be a map or one of the shortcuts 'true' or 'false'", context)
+	}
+
+}
+
+// PlugRule holds the rule of what is allowed, wrt installation and connection, for a plug of a specific interface for a snap.
+type PlugRule struct {
+	Interface string
+
+	AllowInstallation *PlugInstallationConstraints
+	DenyInstallation  *PlugInstallationConstraints
+
+	AllowConnection *PlugConnectionConstraints
+	DenyConnection  *PlugConnectionConstraints
+
+	AllowAutoConnection *PlugConnectionConstraints
+	DenyAutoConnection  *PlugConnectionConstraints
+}
+
+func (r *PlugRule) setPlugInstallationConstraints(field string, cstrs *PlugInstallationConstraints) {
+	switch field {
+	case "allow-installation":
+		r.AllowInstallation = cstrs
+	case "deny-installation":
+		r.DenyInstallation = cstrs
+	default:
+		panic("unknown PlugRule field " + field)
+	}
+}
+
+func (r *PlugRule) setPlugConnectionConstraints(field string, cstrs *PlugConnectionConstraints) {
+	switch field {
+	case "allow-connection":
+		r.AllowConnection = cstrs
+	case "deny-connection":
+		r.DenyConnection = cstrs
+	case "allow-auto-connection":
+		r.AllowAutoConnection = cstrs
+	case "deny-auto-connection":
+		r.DenyAutoConnection = cstrs
+	default:
+		panic("unknown PlugRule field " + field)
+	}
+}
+
+// PlugInstallationConstraints specifies a set of constraints on an interface plug relevant to the installation of snap.
+type PlugInstallationConstraints struct {
+	PlugAttributes *AttributeConstraints
+}
+
+func compilePlugInstallationConstraints(interfaceName string, entry string, constraints interface{}) (*PlugInstallationConstraints, error) {
+	context := fmt.Sprintf("%s in plug rule for interface %q", entry, interfaceName)
+	cMap, shortcut, err := checkMapOrShortcut(context, constraints)
+	if err != nil {
+		return nil, err
+	}
+	if cMap == nil {
+		if shortcut {
+			return &PlugInstallationConstraints{PlugAttributes: AlwaysMatchAttributes}, nil
+		}
+		return &PlugInstallationConstraints{PlugAttributes: NeverMatchAttributes}, nil
+	}
+	attrs, err := compileAttributeConstraints(cMap["plug-attributes"])
```

Comment:
Can we please also introduce plug-snap-type and slot-snap-type on the opposite end? Also lists.